txpool/txpool.go

```diff
@@ -64,126 +76,122 @@ type signer interface {
+// 	and passes validation.
+//
+//	2. When consensus is processing a previously
+// 	promoted transaction and decides to return it
+// 	to the pool (Demote). These requests have the
+// 	demoted flag set to true.
+type enqueueRequest struct {
+	tx      *types.Transaction
+	demoted bool
+}
+
+// A promoteRequest is created each time some account
+// is eligible for promotion. This request is signaled
+// on 2 ocassions:
+//
+// 	1. When an enqueued transaction's nonce is
+// 	not greater than the expected (account's nextNonce).
+// 		== 	nextNonce	- transaction is expected (addTx)
+// 		<	nextNonce	- transaction was demoted (Demote)
+//
+// 	2. When an account's nextNonce is updated (during ResetWithHeader)
+// 	and the first enqueued transaction matches the new nonce.
+type promoteRequest struct {
+	account types.Address
+}
+
+// TxPool is a module that handles pending transactions.
+// All transactions are handled within their respective accounts.
+// An account contains 2 queues a transaction needs to go through:
+// - 1. Enqueued	(entry point)
+// - 2. Promoted	(exit point)
+// 	(both queues are min nonce ordered)
 //
-// There are fundamentally 2 queues in the txpool module:
-// - Account based transactions (accountQueues)
-// - Global valid transactions, from any account (pendingQueue)
+// When consensus needs to process promoted transactions,
+// the pool generates a queue of "executable" transactions. These
+// transactions are the first-in-line of some promoted queue,
+// ready to be written to the state (primaries).
 type TxPool struct {
 	logger     hclog.Logger
 	signer     signer
 	forks      chain.ForksInTime
 	store      store
 	idlePeriod time.Duration
 
-	// Unsorted min heap of transactions per account.
-	// The heap is min nonce based
-	accountQueuesLock sync.Mutex
-	accountQueues     map[types.Address]*accountQueueWrapper
+	// map of all accounts registered by the pool
+	accounts accountsMap
 
-	// Max price heap for all transactions that are valid
-	pendingQueue *txPriceHeap
+	// all the primaries sorted by max gas price
+	executables *pricedQueue
 
-	// Min price heap for all remote transactions
-	remoteTxns *txPriceHeap
+	// lookup map keeping track of all
+	// transactions present in the pool
+	index lookupMap
 
-	// Lookup map that keeps track of txns present in the pool
-	txnLookupMap map[types.Hash]*types.Transaction
-
-	// Lock for the txn lookup map
-	txnLookupMapLock sync.RWMutex
+	// networking stack
+	topic *network.Topic
 
-	// Gauge for measuring pool capacity
+	// gauge for measuring pool capacity
 	gauge slotGauge
 
-	// Networking stack
-	topic *network.Topic
+	// channels on which the pool's event loop
+	// does dispatching/handling requests.
+	enqueueReqCh chan enqueueRequest
+	promoteReqCh chan promoteRequest
 
-	// Flag indicating if the current node is a sealer,
+	// shutdown channel
+	shutdownCh chan struct{}
+
+	// flag indicating if the current node is a sealer,
```

Comment:
I am not entirely sure what was the original reasoning behind this, but this is leftover from the previous implementation. If there is no hard in non-sealers broadcasting transactions, I'll remove the redundant `if` check.

Reply:
After consuting with @zivkovicmilos , non-sealers will never produce blocks, so there is no need for the to accept gossiped transactions. However, transactions added with `AddTx` should still be included, so the other (validator) nodes can hear about them.

Reply:
It's good that we know the cause. In the future this might be a problem if you have a node that's only used for JSON-RPC to query the state of the txpool. Let's create an issue for that for later @dbrajovic